Geometry_Engine/Modify/SplitAtPoints.cs

```diff
@@ -212,111 +212,113 @@ public static List<PolyCurve> SplitAtPoints(this PolyCurve curve, List<Point> po
                 return new List<PolyCurve> { curve.DeepClone() };
 
             List<PolyCurve> result = new List<PolyCurve>();
-            List<ICurve> tmpResult = new List<ICurve>();
-            List<Point> subPoints = new List<Point>();
-            List<Point> onCurvePoints = new List<Point>();
 
-            foreach (Point p in points)
+            List<ICurve> subParts = curve.ISubParts().ToList(); //Subparts of PolyCurve
+            if (subParts.Count == 1)    //Single SubPart
             {
-                if (p.IsOnCurve(curve, tolerance))
-                    onCurvePoints.Add(p);
+                foreach (ICurve splitSegment in SplitAtPoints(subParts[0] as dynamic, points, tolerance))   //Simply call the appropriate split at method
+                {
+                    result.Add(new PolyCurve { Curves = new List<ICurve> { splitSegment } });   //And wrap split segments into polycurves
+                }
+                return result;  
             }
 
-            onCurvePoints = onCurvePoints.CullDuplicates(tolerance);
-
-            if (onCurvePoints.Count == 0)
-                return new List<PolyCurve> { curve.DeepClone() };
-
-            onCurvePoints = onCurvePoints.SortAlongCurve(curve);
-
-            foreach (ICurve crv in curve.SubParts())
+            List<Point> nonDuplicatePoints = points.CullDuplicates(tolerance);
+            PolyCurve prev = null;  //Polycurve to collect parts across segments
+            double sqTol = tolerance * tolerance;
+            for (int k = 0; k < subParts.Count; k++)
             {
-                if (crv is Arc)
+                ICurve crv = subParts[k];
+                List<Point> ptsOnSegment = nonDuplicatePoints.Where(x => x.IIsOnCurve(crv, tolerance)).ToList(); //Get points on the segment
+
+                if (ptsOnSegment.Any()) //If any points on the current segment
                 {
-                    foreach (Point point in onCurvePoints)
+                    List<ICurve> split = new List<ICurve>();
+                    split.AddRange(SplitAtPoints(crv as dynamic, ptsOnSegment, tolerance)); //Split the segment. Assuming curves returned are ordered
```

Comment:
Couldn't you simply call `ISplitAtPoints` here?
```suggestion
                    List<ICurve> split = SplitAtPoints(crv as dynamic, ptsOnSegment, tolerance); //Split the segment. Assuming curves returned are ordered
```